// @sap/approuter/lib/backend-request/agents.js
'use strict';

const Cache = require('lru-cache');
const { HttpProxyAgent } = require('http-proxy-agent');
const { HttpsProxyAgent } = require('https-proxy-agent');
const { Agent: HttpAgent } = require('http');
const { Agent: HttpsAgent } = require('https');
const HttpKeepAliveAgent = require('agentkeepalive');
const HttpsKeepAliveAgent = HttpKeepAliveAgent.HttpsAgent;
const isConReuseDisabled = isConnectionReuseDisabled();

// http-proxy-agent v7 builds the proxy request-line URL from req.getHeader('host').
// For OnPremise destinations with preserveHostHeader, the Host header is the original
// client host, but the request-line must use the destination virtual hostname (for
// Cloud Connector ACL matching). This subclass pre-sets req.path to an absolute URL
// in addRequest, before the parent's setRequestProps runs. Since new URL(absoluteUrl, base)
// ignores the base per WHATWG URL spec, the request-line preserves the virtual host
// while the Host header remains the original — restoring v4 behavior.
class OnPremiseHttpProxyAgent extends HttpProxyAgent {
  addRequest(req, opts) {
    const hostname = opts.hostname || opts.host || 'localhost';
    const port = opts.port && opts.port !== 80 ? ':' + opts.port : '';
    req.path = 'http://' + hostname + port + req.path;
    super.addRequest(req, opts);
  }
}

// STRICT SETTINGS (mTLS/Certificate traffic)
const STRICT_OPTIONS = {
  keepAlive: true,
  freeSocketTimeout: 4000,
  maxSockets: 200,
  maxFreeSockets: 20,
  scheduling: 'fifo',
  keepAliveMsecs: 1000,
  maxCachedSessions: 0,
  secureOptions: require('crypto').constants.SSL_OP_NO_TICKET
};

function isRunningOnCF() {
  try {
    const vcapApp = JSON.parse(process.env.VCAP_APPLICATION || '{}');
    return !!vcapApp.cf_api;
  } catch (_e) {
    return false;
  }
}

function isConnectionReuseDisabled() {
  const disableValue = process.env.DISABLE_CONNECTION_REUSE;
  if (disableValue === 'true') {
    return true;
  } else if (disableValue === 'false') {
    return false;
  } else if (!disableValue && !isRunningOnCF()) {
    return true;
  }
  return false;
}

if (isConReuseDisabled) {
  exports.proxyAgentsCache = new Cache(1);
} else {
  exports.proxyAgentsCache = new Cache(10);
}
exports.httpAgent = new HttpKeepAliveAgent();
exports.httpsAgent = new HttpsKeepAliveAgent();
exports.httpStrictAgent = new HttpKeepAliveAgent(STRICT_OPTIONS);
exports.httpsStrictAgent = new HttpsKeepAliveAgent(STRICT_OPTIONS);


function isSecureConnection(protocol) {
  return protocol === 'https:' || protocol === 'wss:';
}

exports.get = function (protocol, proxy, destination) {
  const connectivityScenario = destination && destination.proxyType === 'OnPremise';
  let isSecure = connectivityScenario ? false : isSecureConnection(protocol);
  if (proxy) {
    if (connectivityScenario && destination.preserveHostHeader) {
      return retrieveOnPremPreserveHostProxyAgent(proxy);
    }
    return retrieveProxyAgent(proxy, isSecure);
  }
  if (isConReuseDisabled) {
    return isSecure ? new HttpsAgent() : new HttpAgent();
  }
  if (destination && destination.strictConnectionMode) {
    return isSecure ? exports.httpsStrictAgent : exports.httpStrictAgent;
  }
  return isSecure ? exports.httpsAgent : exports.httpAgent;
};

function retrieveProxyAgent(proxy, isSecure) {
  let id = buildIdentifier(proxy, isSecure);
  let agent = exports.proxyAgentsCache.get(id);
  if (agent) {
    return agent;
  }
  agent = isSecure ? new HttpsProxyAgent(proxy) : new HttpProxyAgent(proxy);
  exports.proxyAgentsCache.set(id, agent);
  return agent;
}

function retrieveOnPremPreserveHostProxyAgent(proxy) {
  let id = proxy + ':onpremise';
  let agent = exports.proxyAgentsCache.get(id);
  if (agent) {
    return agent;
  }
  agent = new OnPremiseHttpProxyAgent(proxy);
  exports.proxyAgentsCache.set(id, agent);
  return agent;
}

function buildIdentifier(proxy, isSecure) {
  let suffix = isSecure ? 'secure' : 'plain';
  return proxy + ':' + suffix;
}
