// @sap/approuter/lib/backend-request/memstore.js
'use strict';


/*
 * express-session
 * Copyright(c) 2010 Sencha Inc.
 * Copyright(c) 2011 TJ Holowaychuk
 * Copyright(c) 2015 Douglas Christopher Wilson
 * MIT Licensed
 */

/**
 * Module dependencies.
 * @private
 */

const session = require('express-session');
const util = require('util');
const tracer = require('../utils/logger').getTracer(__filename);
const cookieUtils = require('../utils/cookie-utils');

/**
 * Shim setImmediate for node.js < 0.10
 * @private
 */

/**
 * A session store in memory.

 * @param {Object} [options]
 * @param {Number} [options.timeout] Session timeout (in minutes). Number has to be positive.
 * @public
 */

let Store = session.Store;

function MemoryStore(options) {
  Store.call(this);
  options = options || {};
  this.store = options.externalSessionStore;
  this.options = {};
  this.sessions = {};
  this.sessionTimers = {};
  this.pendingRedisFetches = {}; // Track pending Redis fetch promises to avoid race conditions
  this.inFlightSessions = new Set(); // Track sessionIds currently being processed (no-stickiness only)
  if (options.timeout && options.timeout < 1) {
    options.timeout = null;
  }

  this.options.timeout = (options.timeout || 15) * 6;

  let getSessionFn = getSession.bind(this);
  let self = this;
  this.sessionChecker = setInterval(function () {
    for (let sessionId in self.sessionTimers) {
      if (--self.sessionTimers[sessionId] <= 0) {
        getSessionFn(sessionId, async (sessionObject, err) => {
          if (!err && sessionObject) {
            // In no-stickiness mode, another instance may have reset the Redis TTL
            // since our local timer started counting down. Check Redis before firing
            // timeout to avoid prematurely destroying an active session.
            if (!cookieUtils.isSessionStickinessEnabled() && self.store) {
              try {
                const existsInRedis = await self.store.exists(sessionId);
                if (existsInRedis) {
                  // Session is still alive in Redis — reset local timer and skip timeout
                  self.sessionTimers[sessionId] = self.options.timeout;
                  return;
                }
              } catch (redisErr) {
                // Redis unavailable - proceed with local timeout rather than crash
                if (tracer.isEnabled('warning')) {
                  tracer.warning('Redis check failed during session timeout for %s: %s', sessionId, redisErr.message);
                }
              }
            }
            self.emit('timeout', sessionObject);
            self.destroy(sessionId);
          }
        });
      }
    }
  },
  10 * 1000); // runs every 10 seconds
  return this;
}

/**
 * Inherit from Store.
 *
 */

util.inherits(MemoryStore, Store);

/**
 * Called by express-session after loading a session, giving us access to req/res.
 * In no-stickiness mode we attach a one-time 'finish' listener so the in-flight
 * flag is always cleared at end-of-request — even when express-session skips set()
 * because the session was not modified (resave:false, saveUninitialized:false).
 */
MemoryStore.prototype.createSession = function createSession(req, session) {
  // Call the base implementation first
  Store.prototype.createSession.call(this, req, session);

  if (!cookieUtils.isSessionStickinessEnabled() && req && req.res) {
    const sessionId = req.sessionID;
    const store = this;
    // Clear in-flight on both finish (normal end) and close (abrupt TCP drop).
    // Using a named handler so we can remove the other listener once one fires,
    // preventing a double-delete or a dangling listener on the response object.
    function clearInFlight() {
      store.inFlightSessions.delete(sessionId);
      req.res.removeListener('finish', clearInFlight);
      req.res.removeListener('close', clearInFlight);
    }
    req.res.once('finish', clearInFlight);
    req.res.once('close', clearInFlight);
  }

  return req.session;
};

/**
 * Get all active sessions.
 *
 * @param {function} callback
 * @public
 */

MemoryStore.prototype.all = function all(callback) {
  let sessions = {};
  Object.entries(this.sessions).forEach(([ sessionId, session ]) => {
    const sessionObject = JSON.parse(session);
    sessionObject.id = sessionId;
    sessions[sessionId] = JSON.stringify(sessionObject);
  });
  callback && callback(null, sessions);
};

/**
 * Clear all sessions.
 *
 * @param {function} callback
 * @public
 */

MemoryStore.prototype.clear = function clear(callback) {
  this.store && this.store.clear();
  this.sessions = {};
  this.sessionTimers = {};
  callback && callback();
};

/**
 * Destroy the session associated with the given IAS session ID.
 *
 * @param {string} sid
 * @public
 */

MemoryStore.prototype.destroyIASSession = async function (sid) {
  let sessionId = Object.keys(this.sessions).find(sessionId => {
    let session = JSON.parse(this.sessions[sessionId]);
    return session.user && session.user.sid === sid;
  });
  if (sessionId) {
    this.destroy(sessionId);
    return true;
  } else {
    return false;
  }
};

/**
 * Destroy the session associated with the given session ID.
 *
 * @param {string} sessionId
 * @public
 */

MemoryStore.prototype.destroy = function destroy(sessionId, callback) {
  // Clear in-flight flag so regenerate() starts fresh with the new session ID
  this.inFlightSessions.delete(sessionId);

  // Destroy session in Redis first
  if (this.store) {
    this.store.destroy(sessionId);
  }

  if (!this.sessions[sessionId]) {
    return callback && callback();
  }

  let session = this.sessions[sessionId];
  delete this.sessionTimers[sessionId];
  delete this.sessions[sessionId];
  this.emit('destroy', JSON.parse(session));
  callback && callback();
};

/**
 * Fetch session by the given session ID. Never returns error.
 *
 * @param {string} sessionId
 * @param {function} callback
 * @public
 */

MemoryStore.prototype.get = function get(sessionId, callback) {
  // When session cookie name is not JSESSIONID, CF session stickiness doesn't work
  // In this case, always fetch from Redis to update the in-memory session
  const shouldAlwaysFetchFromRedis = Boolean(!cookieUtils.isSessionStickinessEnabled()) && this.store;

  if (shouldAlwaysFetchFromRedis) {
    getSessionFromRedisAndUpdate.call(this, sessionId, session => callback(null, session));
  } else {
    getSessionResetTimeout.call(this, sessionId, session => callback(null, session));
  }
};

/**
 * Check if a session exists in the store without parsing it.
 *
 * @param {string} sessionId
 * @returns {boolean}
 * @public
 */
MemoryStore.prototype.exists = function exists(sessionId) {
  return this.sessions.hasOwnProperty(sessionId);
};

/**
 * Commit the given session associated with the given sessionId to the store.
 *
 * @param {string} sessionId
 * @param {object} session
 * @param {function} callback
 * @public
 */

/**
 * Get number of active sessions.
 *
 * @param {function} callback
 * @public
 */

MemoryStore.prototype.length = function length(callback) {
  callback(null, Object.keys(this.sessions).length);
};

MemoryStore.prototype.set = function set(sessionId, session, callback) {

  const useVersioning =  Boolean(!cookieUtils.isSessionStickinessEnabled());

  // For non-versioning case (default JSESSIONID), update synchronously for backward compatibility
  if (!useVersioning) {
    const stringSession = JSON.stringify(session, function (key, value) {
      return (key === 'id') ? undefined : value;
    });

    // Update in-memory cache synchronously
    let isUpdate = this.sessions.hasOwnProperty(sessionId);
    this.sessions[sessionId] = stringSession;
    if (isUpdate) {
      this.emit('update', sessionId);
    }

    // Reset timer synchronously
    resetTimer.call(this, sessionId);

    // Store in Redis asynchronously (fire-and-forget)
    if (this.store) {
      this.store.set(sessionId, stringSession, this.options.timeout / 6).catch(err => {
        if (tracer.isEnabled('Error')) {
          tracer.error('Failed to store session in Redis: %s', err.message);
        }
      });
    }

    callback && callback();
    return;
  }

  // For versioning case, use async storeSession with optimistic locking.
  // This is the express-session save path (end of request) — the only write
  // where CAS conflict detection is appropriate.
  storeSession.call(this, sessionId, session, true).then(result => {
    // Clear in-flight flag — the final session state is now in Redis and
    // the next request for this session should fetch fresh from Redis.
    this.inFlightSessions.delete(sessionId);
    if (result === false) {
      // Version conflict: another instance wrote a newer session version to Redis.
      // No retry is performed intentionally — retrying would risk overwriting the
      // other instance's fresher data. express-session receives this error via the
      // callback and calls next(err), so the request fails visibly and the client
      // can retry, at which point it will re-read the updated session from Redis.
      const error = new Error('Session version conflict - session was modified concurrently');
      error.code = 'EVERSION';
      callback && callback(error);
    } else {
      // CAS script already set the TTL in Redis via PX — only update local timer.
      resetTimer.call(this, sessionId, true);
      callback && callback();
    }
  }).catch(err => {
    callback && callback(err);
  });
};

/**
 * Touch the given session object associated with the given session ID.
 *
 * @param {string} sessionId
 * @param {object} session
 * @param {function} callback
 * @public
 */

MemoryStore.prototype.touch = function touch(sessionId, session, callback) {
  getSessionResetTimeout.call(this, sessionId, currentSession => {
    if (currentSession) {
      // update expiration
      currentSession.cookie = session.cookie;
      storeSession.call(this, sessionId, currentSession).then(() => {
        callback && callback();
      }).catch(err => {
        callback && callback(err);
      });
    } else {
      callback && callback();
    }
  });
};

/**
 * Reset timer for session
 *
 * @param {string} sessionId
 * @param {function} callback
 * @public
 */

MemoryStore.prototype.resetSessionTimer = function (sessionId, callback) {
  let error = null;
  if (this.sessionTimers[sessionId]) {
    resetTimer.call(this, sessionId);
  } else {
    error = new Error('Could not reset session timer! Session ' + sessionId + ' missing!');
  }
  return callback && callback(error);
};


/**
 * Get session timeout
 * @returns {Number} Default session timeout in minutes
 * @public
 */

MemoryStore.prototype.getDefaultSessionTimeout = function () {
  return this.options.timeout / 6;
};

/**
 * Get specific session timer
 * @param sessionId
 * @param callback callback(err, time), time is in minutes
 * @public
 */

MemoryStore.prototype.getSessionTimeout = function (sessionId, callback) {
  let timer = this.sessionTimers[sessionId];
  callback(timer ? null : new Error('Invalid session'), timer && timer / 6);
};

/**
 * Updates the session in the memstore by applying function change to it without changing the session timer.
 *
 * @param {string} sessionId
 * @param {function} change A function that changes the session - function change(session)
 * @param {boolean} shouldResetTimer defaults to true
 * @param {boolean} flushToRedis when true, writes through to Redis immediately even if in-flight
 *                  — used for cross-instance coordination (e.g. jwtRefreshStarted).
 */
MemoryStore.prototype.update = function (sessionId, change, shouldResetTimer, flushToRedis) {
  getSession.call(this, sessionId, (session, err) => {
    if (err && tracer.isEnabled('Error')) {
      tracer.error('Failed to get session with sessionId %s', sessionId);
    }
    if (!session) {
      return change(null);
    }
    change(session);
    storeSession.call(this, sessionId, session, false, flushToRedis).then(result => {
      if (result === false && tracer.isEnabled('warning')) {
        tracer.warning('Failed to update session %s due to version conflict', sessionId);
      }
      if (shouldResetTimer !== false && result !== false) {
        // If flushToRedis, the SET already included the TTL — only update local timer.
        resetTimer.call(this, sessionId, flushToRedis);
      }
    }).catch(err => {
      if (tracer.isEnabled('Error')) {
        tracer.error('Failed to store session %s during update: %s', sessionId, err.message);
      }
    });
  });
};

MemoryStore.prototype.getSessionWithoutResetTimeout = function (sessionId, callback) {
  return getSession.call(this, sessionId, callback);
};

/**
 * Get session from the store and reset timeout.
 * @param {string} sessionId
 * @param callback
 * @returns {object} session
 * @private
 */
function getSessionResetTimeout(sessionId, callback) {
  getSession.call(this, sessionId, (session, err) => {
    if (err && tracer.isEnabled('Error')) {
      tracer.error('Failed to get session with sessionId %s', sessionId);
    }
    if (session && !session.externalSessionId) {
      resetTimer.call(this, sessionId);
    }
    callback(session);
  });
}

/**
 * Gets the session from the remote store
 * @param sessionId
 */
async function getSessionFromRemoteStore(sessionId) {
  const session = await this.store.get(sessionId);
  if (session) {
    // RedisStore.get() already returns a parsed object (decompress calls JSON.parse internally)
    const sessionObject = typeof session === 'string' ? JSON.parse(session) : session;
    sessionObject.id = sessionId;
    // In sticky mode only: reset jwtRefreshStarted so the flag doesn't get stuck.
    // In no-stickiness mode we preserve the value from Redis — it was written there
    // by another instance to signal that a refresh is already in progress, and
    // resetting it here would allow every instance to start a duplicate refresh.
    if (cookieUtils.isSessionStickinessEnabled()) {
      sessionObject.jwtRefreshStarted = false;
    }
    return sessionObject;
  }
}

/**
 * Fetch session from Redis and update in-memory session.
 * Used when session stickiness is not available (custom cookie name != JSESSIONID)
 * Prevents race conditions by ensuring only one Redis fetch per session at a time.
 * @param {string} sessionId
 * @param callback
 * @private
 */
function getSessionFromRedisAndUpdate(sessionId, callback) {
  // If this session is already in-flight (loaded from Redis for this request),
  // use the in-memory data directly — skip the Redis fetch to preserve local changes.
  if (this.inFlightSessions.has(sessionId)) {
    getSession.call(this, sessionId, session => callback(session));
    return;
  }

  // Check if there's already a pending fetch for this session
  if (this.pendingRedisFetches[sessionId]) {
    // Wait for the existing fetch to complete
    this.pendingRedisFetches[sessionId].then(session => {
      try {
        callback(session);
      } catch (err) {
        if (tracer.isEnabled('Error')) {
          tracer.error('Error in callback while waiting for session %s: %s', sessionId, err.message);
        }
      }
    }).catch(() => {
      try {
        callback();
      } catch (err) {
        if (tracer.isEnabled('Error')) {
          tracer.error('Error in callback error handler for session %s: %s', sessionId, err.message);
        }
      }
    });
    return;
  }

  // Create a new fetch promise
  if (tracer.isEnabled('debug')) {
    tracer.debug('Non-sticky mode: fetching session %s from Redis', sessionId);
  }
  const fetchPromise = getSessionFromRemoteStore.call(this, sessionId)
    .then(async session => {
      if (tracer.isEnabled('debug')) {
        tracer.debug('Session %s %s in Redis', sessionId, session ? 'found' : 'not found');
      }
      if (session) {
        // Mark as in-flight BEFORE storeSession so the guard in storeSession
        // correctly skips the intermediate Redis write on the initial fetch write-back.
        this.inFlightSessions.add(sessionId);
        await storeSession.call(this, sessionId, session);
        // Session was just read from Redis — it exists, no need for EXISTS check.
        // Reset TTL directly and update local timer.
        resetTimer.call(this, sessionId, false, true);
      } else {
        // No session in Redis - clear stale in-memory session to stay in sync
        delete this.sessions[sessionId];
        delete this.sessionTimers[sessionId];
      }
      return session;
    })
    .catch(err => {
      if (tracer.isEnabled('Error')) {
        tracer.error('Failed to fetch session from Redis for sessionId %s: %s', sessionId, err.message);
      }
      return null;
    })
    .finally(() => {
      // Clean up the pending fetch tracking
      delete this.pendingRedisFetches[sessionId];
    });

  // Store the pending fetch promise
  this.pendingRedisFetches[sessionId] = fetchPromise;

  // Wait for fetch to complete and call callback
  fetchPromise.then(session => {
    callback(session);
  }).catch(err => {
    if (tracer.isEnabled('Error')) {
      tracer.error('Error in callback for session %s: %s', sessionId, err.message);
    }
    callback();
  });
}

/**
 * Get session from the store.
 * @param {string} sessionId
 * @param callback
 * @returns {object} session
 * @private
 */
function getSession(sessionId, callback) {
  const sessionJson = this.sessions[sessionId];
  if (sessionJson) {
    const session = JSON.parse(sessionJson);
    Object.defineProperty(session, 'id', { value: sessionId });
    callback(session);
  } else if (this.store) {
    getSessionFromRemoteStore.call(this, sessionId).then(callback);
  } else {
    callback();
  }
}

async function resetTimerRemoteStore(sessionId) {
  const exists = await this.store.exists(sessionId);
  if (exists) {
    await this.store.resetTimer(sessionId, this.options.timeout / 6);
  }
}

async function resetTimerRemoteStoreNoCheck(sessionId) {
  await this.store.resetTimer(sessionId, this.options.timeout / 6);
}

/**
 * Reset timer for session
 * @param {string} sessionId
 * @param {boolean} skipRedis  when true, skip Redis TTL update (caller already set TTL via SET/EVAL)
 * @param {boolean} skipExists when true, skip the EXISTS check before EXPIRE (key is known to exist)
 * @private
 */
function resetTimer(sessionId, skipRedis, skipExists) {
  if (this.store && !skipRedis) {
    if (skipExists) {
      resetTimerRemoteStoreNoCheck.call(this, sessionId);
    } else {
      resetTimerRemoteStore.call(this, sessionId);
    }
  }
  let isUpdate = this.sessionTimers.hasOwnProperty(sessionId);
  this.sessionTimers[sessionId] = this.options.timeout;
  if (isUpdate) {
    this.emit('update', sessionId, this.options.timeout / 6); // in minutes
  }
}

async function storeSession(sessionId, session, useOptimisticLocking = false, flushToRedis = false) {
  const stringSession = JSON.stringify(session, function (key, value) {
    return (key === 'id') ? undefined : value; // skip id
  });

  const useVersioning = Boolean(!cookieUtils.isSessionStickinessEnabled());

  // CAS is only used when the caller explicitly requests it (MemoryStore.prototype.set).
  // Internal paths (update, touch, getSessionFromRedisAndUpdate) write unconditionally —
  // they operate on already-loaded session data and must not be blocked by CAS conflicts.
  if (this.store && useVersioning && useOptimisticLocking) {
    // Pass the current in-memory string as the expected value for CAS.
    // MemoryStore.prototype.set is only called by express-session at request end, after
    // getSessionFromRedisAndUpdate has already synced in-memory from Redis, so this
    // is the correct CAS baseline — no extra GET round-trip is needed.
    const expectedValue = this.sessions[sessionId] || null;
    const result = await this.store.set(sessionId, stringSession, this.options.timeout / 6, true, expectedValue);
    if (result === false) {
      if (tracer.isEnabled('warning')) {
        tracer.warning('Version conflict detected for session %s - session was modified concurrently', sessionId);
      }
      // Don't update in-memory cache if Redis write failed due to conflict
      return false;
    }
  } else if (this.store && useVersioning) {
    // Non-CAS write for internal paths (mid-request): skip Redis if session is in-flight,
    // unless flushToRedis=true (used for cross-instance coordination like jwtRefreshStarted).
    if (flushToRedis || !this.inFlightSessions.has(sessionId)) {
      this.store.set(sessionId, stringSession, this.options.timeout / 6).catch(err => {
        if (tracer.isEnabled('Error')) {
          tracer.error('Failed to store session in Redis: %s', err.message);
        }
      });
    }
  }

  // Update in-memory cache (happens synchronously for backward compatibility)
  let isUpdate = this.sessions.hasOwnProperty(sessionId);
  this.sessions[sessionId] = stringSession;
  if (isUpdate) {
    this.emit('update', sessionId);
  }

  // If not using versioning, store in Redis asynchronously (fire-and-forget for backward compatibility)
  if (this.store && !useVersioning) {
    this.store.set(sessionId, stringSession, this.options.timeout / 6).catch(err => {
      if (tracer.isEnabled('Error')) {
        tracer.error('Failed to store session in Redis: %s', err.message);
      }
    });
  }

  return true;
}

module.exports = MemoryStore;
