// @sap/approuter/lib/passport/state.js
'use strict';

const NodeCache = require('node-cache');
const {getValidatedEnvInt} = require('../utils/env-utils');
const cookieUtils = require('../utils/cookie-utils');
const loggerUtils = require('../utils/logger');
const logger = loggerUtils.getLogger('/Auth/state');
let _logRequestInfo;
function logRequestInfo(req, message) {
  if (!_logRequestInfo) {
    _logRequestInfo = require('../utils/application-logs-utils').logRequestInfo;
  }
  _logRequestInfo(req, message);
}

const stdTTL              = getValidatedEnvInt('STATE_CACHE', 180);
const CHECK_PERIOD         = stdTTL + 20;
const STATE_CACHE          = new NodeCache({stdTTL: stdTTL, checkperiod: CHECK_PERIOD});

const STATE_PREFIX = '{approuter-state}';

/**
 * Returns the state cookie name based on session stickiness:
 * - Session stickiness (JSESSIONID): state cookie name = session cookie name
 * - No session stickiness (custom name): state cookie name = <session cookie name>_STATE
 */
function getStateCookieName() {
  const sessionCookieName = cookieUtils.getSessionCookieName();
  if (cookieUtils.isSessionStickinessEnabled()) {
    return sessionCookieName;
  }
  return sessionCookieName + '_STATE';
}

function getRedisStore(req) {
  const store = req && req.app && typeof req.app.get === 'function' && req.app.get('externalSessionStore');
  return (store && typeof store.checkClient === 'function') ? store : null;
}

// --- Redis helpers ---

async function redisSet(key, value, req) {
  const redisStore = getRedisStore(req);
  if (!redisStore) {
    return false;
  }
  if (!redisStore.checkClient()) {
    if (!cookieUtils.isSessionStickinessEnabled()) {
      logger.warn(`OAuth state store: Redis client not ready — falling back to local cache. In non-sticky deployments this state entry will not be visible to other instances and the OAuth callback will fail with 401. [key=${key}]`);
    }
    return false;
  }
  await redisStore.updateExternalSession({
    externalSessionId: key,
    externalSessionExpiration: (Date.now() + stdTTL * 1000) / 1000,
    value
  }, STATE_PREFIX);
  return true;
}

async function redisGet(key, req) {
  const redisStore = getRedisStore(req);
  if (!redisStore || !redisStore.checkClient()) {
    return undefined;
  }
  const found = await redisStore.getExternalSession(key, null, STATE_PREFIX);
  return found ? found.value : null;
}

async function redisDel(key, req) {
  const redisStore = getRedisStore(req);
  if (!redisStore || !redisStore.checkClient()) {
    return false;
  }
  await redisStore.destroyExternalSession(key, STATE_PREFIX);
  return true;
}

// --- Public API ---

async function setCacheStateValue(stateValue, req) {
  logRequestInfo(req, `Storing OAuth state [stateValue=${stateValue}]`);
  if (await redisSet(stateValue, stateValue, req)) {
    logRequestInfo(req, `OAuth state stored in Redis [stateValue=${stateValue}]`);
    return;
  }
  STATE_CACHE.set(stateValue, stateValue);
}

async function getCacheStateValue(stateValue, req) {
  const redisValue = await redisGet(stateValue, req);
  if (redisValue !== undefined) {
    const match = redisValue === stateValue;
    logRequestInfo(req, `OAuth state lookup in Redis: ${match ? 'matched' : 'not found'} [stateValue=${stateValue}]`);
    return match ? stateValue : null;
  }
  const cached = STATE_CACHE.get(stateValue);
  logRequestInfo(req, `OAuth state lookup in local cache: ${cached ? 'found' : 'not found'} [stateValue=${stateValue}]`);
  return cached;
}

function deleteCacheStateValue(value) {
  if (!value) {
    return;
  }
  STATE_CACHE.del(value);
}

/**
 * Store mapping from state parameter to session cookie value.
 * Redis key: state-cookie:${state}  (mirrors local cache key)
 *
 * @param {string} state - The state parameter value (public, in URL)
 * @param {string} cookieValue - The session cookie value (secret, HttpOnly)
 * @param {object} [req] - The request object
 */
async function setStateToCookieMapping(state, cookieValue, req) {
  if (!state || !cookieValue) {
    throw new Error('State and cookie value are required');
  }
  if (await redisSet(`state-cookie:${state}`, cookieValue, req)) {
    logRequestInfo(req, `State-to-cookie mapping stored in Redis [stateValue=${state}]`);
    return;
  }
  STATE_CACHE.set(`state-cookie:${state}`, cookieValue);
}

/**
 * Retrieve the expected cookie value for a given state parameter.
 * Redis key: state-cookie:${state}
 *
 * @param {string} state - The state parameter value
 * @param {object} [req] - The request object
 * @returns {Promise<string|null|undefined>} The cookie value, or null/undefined if not found
 */
async function getStateToCookieMapping(state, req) {
  if (!state) {
    return undefined;
  }
  const redisValue = await redisGet(`state-cookie:${state}`, req);
  if (redisValue !== undefined) {
    logRequestInfo(req, `State-to-cookie mapping lookup in Redis: ${redisValue ? 'found' : 'not found'} [stateValue=${state}]`);
    return redisValue;
  }
  return STATE_CACHE.get(`state-cookie:${state}`);
}

/**
 * Delete state-to-cookie mapping after use.
 * Redis key: state-cookie:${state}
 *
 * @param {string} state - The state parameter value
 * @param {object} [req] - The request object
 */
async function deleteStateToCookieMapping(state, req) {
  if (!state) {
    return;
  }
  if (await redisDel(`state-cookie:${state}`, req)) {
    return;
  }
  STATE_CACHE.del(`state-cookie:${state}`);
}

/**
 * Check if a cookie value corresponds to an active OAuth flow.
 * Used to filter out pre-auth session cookies from stale cookie cleanup.
 *
 * @param {string} cookie - Cookie string in format "name=value"
 * @returns {boolean} True if cookie is associated with an active OAuth flow
 */
function isActiveStateCookie(cookie) {
  try {
    const [name, value] = cookie.trim().split('=');
    // In sticky mode the state cookie name equals the session cookie name (JSESSIONID).
    // Any cookie with that name is a live session/state cookie — value lookup is not
    // meaningful here since the same name is used for both session and state.
    if (cookieUtils.isSessionStickinessEnabled() && name.trim() === getStateCookieName()) {
      return true;
    }
    return STATE_CACHE.has(value);
  } catch (_) {
    return false;
  }
}

/**
 * Store PKCE code verifier associated with session cookie value.
 * Redis key: pkce:${sessionId}  (mirrors local cache key)
 * The verifier is keyed by the COOKIE value (secret), not the state (public).
 *
 * @param {string} sessionId - The session cookie value (secret)
 * @param {string} codeVerifier - The PKCE code verifier to store
 * @param {object} [req] - The request object
 * @throws {Error} If sessionId or codeVerifier is missing
 */
async function setCodeVerifier(sessionId, codeVerifier, req) {
  if (!sessionId || !codeVerifier) {
    throw new Error('Session ID and code verifier are required');
  }
  if (await redisSet(`pkce:${sessionId}`, codeVerifier, req)) {
    logRequestInfo(req, `PKCE code verifier stored in Redis [sessionId=${sessionId}]`);
    return;
  }
  STATE_CACHE.set(`pkce:${sessionId}`, codeVerifier);
}

/**
 * Retrieve PKCE code verifier for session cookie value.
 * Redis key: pkce:${sessionId}
 *
 * @param {string} sessionId - The session cookie value
 * @param {object} [req] - The request object
 * @returns {Promise<string|null|undefined>} The code verifier, or null/undefined if not found
 */
async function getCodeVerifier(sessionId, req) {
  if (!sessionId) {
    return undefined;
  }
  const redisValue = await redisGet(`pkce:${sessionId}`, req);
  if (redisValue !== undefined) {
    logRequestInfo(req, `PKCE code verifier lookup in Redis: ${redisValue ? 'found' : 'not found'} [sessionId=${sessionId}]`);
    return redisValue;
  }
  return STATE_CACHE.get(`pkce:${sessionId}`);
}

/**
 * Delete PKCE code verifier after use.
 * Redis key: pkce:${sessionId}
 *
 * @param {string} sessionId - The session cookie value
 * @param {object} [req] - The request object
 */
async function deleteCodeVerifier(sessionId, req) {
  if (!sessionId) {
    return;
  }
  if (await redisDel(`pkce:${sessionId}`, req)) {
    return;
  }
  STATE_CACHE.del(`pkce:${sessionId}`);
}

module.exports = {
  setCacheStateValue,
  getCacheStateValue,
  getStateCookieName,
  deleteCacheStateValue,
  isActiveStateCookie,
  setStateToCookieMapping,
  getStateToCookieMapping,
  deleteStateToCookieMapping,
  setCodeVerifier,
  getCodeVerifier,
  deleteCodeVerifier,
  getStateTtl: () => stdTTL
};
