// @sap-ux/eslint-plugin-fiori-tools/lib/project-context/parser/service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildAnnotationIndexKey = buildAnnotationIndexKey;
exports.buildServiceIndex = buildServiceIndex;
const odata_annotation_core_1 = require("@sap-ux/odata-annotation-core");
const constants_1 = require("../../constants");
/**
 * Builds an annotation index key from target and term.
 *
 * @param target - The annotation target
 * @param term - The annotation term
 * @returns The index key in format "target/@term"
 */
function buildAnnotationIndexKey(target, term) {
    return `${target}/@${term}`;
}
/**
 * Indexes annotations by their annotation path (target and term).
 * For OData V2 services, also indexes synthetic entries from inline sap:text/sap:label attributes.
 *
 * @param service - Service artifacts containing annotation files
 * @param v2Annotations - Pre-parsed V2 annotations from convertMetadataDocumentV2
 * @returns Annotation index organized by path and qualifier
 */
function indexAnnotationsByAnnotationPath(service, v2Annotations = []) {
    const index = {};
    for (const file of service.fileSequence) {
        const annotationFile = service.annotationFiles[file];
        if (!annotationFile) {
            continue;
        }
        const namespace = annotationFile.namespace ?? { name: 'DEFAULT_NAMESPACE', type: 'namespace' };
        const namespaces = (0, odata_annotation_core_1.getAllNamespacesAndReferences)(namespace, annotationFile.references);
        const aliasInfo = (0, odata_annotation_core_1.getAliasInformation)(namespaces, service.metadataService.getNamespaces());
        for (const target of annotationFile.targets) {
            const targetName = (0, odata_annotation_core_1.toFullyQualifiedName)(aliasInfo.aliasMap, namespace.name, (0, odata_annotation_core_1.parseIdentifier)(target.name));
            if (!targetName) {
                continue;
            }
            processTargetAnnotations(target, annotationFile, namespace, aliasInfo, targetName, index);
        }
    }
    processTargetV2Annotations(v2Annotations, index);
    return index;
}
/**
 * Processes annotations for a specific target and adds them to the index.
 *
 * @param target - The annotation target to process
 * @param annotationFile - The annotation file containing the target
 * @param namespace - Namespace information
 * @param aliasInfo - Alias information for resolving qualified names
 * @param targetName - Fully qualified target name
 * @param index - Annotation index to update
 */
function processTargetAnnotations(target, annotationFile, namespace, aliasInfo, targetName, index) {
    for (const annotation of target.terms) {
        const term = (0, odata_annotation_core_1.getElementAttributeValue)(annotation, "Term" /* Edm.Term */);
        const qualifier = (0, odata_annotation_core_1.getElementAttribute)(annotation, "Qualifier" /* Edm.Qualifier */);
        const termName = (0, odata_annotation_core_1.toFullyQualifiedName)(aliasInfo.aliasMap, namespace.name, (0, odata_annotation_core_1.parseIdentifier)(term));
        if (!termName) {
            continue;
        }
        const termKey = buildAnnotationIndexKey(targetName, termName);
        index[termKey] ??= {};
        const qualifierKey = qualifier?.value ?? 'undefined';
        const indexedValue = index[termKey][qualifierKey];
        const reference = {
            uri: annotationFile.uri,
            value: annotation
        };
        if (indexedValue) {
            indexedValue.top = reference;
            indexedValue.layers.unshift(reference);
        }
        else {
            index[termKey][qualifierKey] = {
                source: annotationFile.uri,
                target: targetName,
                term: termName,
                qualifier: qualifier?.value,
                top: reference,
                layers: [reference]
            };
        }
    }
}
/**
 * Creates a minimal synthetic Element node carrying a single string attribute.
 * Attaches an ESLint-compatible `loc` object when source range is available,
 * enabling ESLint to report at the correct line/column in the metadata file.
 *
 * @param attrName - Attribute name on the Annotation element (e.g. "Path" or "String")
 * @param attrValue - Attribute value
 * @param sourceRange - Optional source range (0-indexed) used for ESLint reporting
 * @returns A synthetic Element, optionally with position range
 */
function createSyntheticElement(attrName, attrValue, sourceRange) {
    const element = (0, odata_annotation_core_1.createElementNode)({
        name: "Annotation" /* Edm.Annotation */,
        attributes: {
            [attrName]: (0, odata_annotation_core_1.createAttributeNode)(attrName, attrValue)
        },
        content: []
    });
    if (sourceRange) {
        element.range = sourceRange;
    }
    return element;
}
/**
 * Indexes synthetic Common.Text and Common.Label annotation entries derived from OData V2
 * inline `sap:text` and `sap:label` attributes, using pre-parsed V2Annotation objects.
 *
 * Only adds entries when no explicit vocabulary annotation already exists for that key.
 *
 * @param v2Annotations - Pre-parsed V2 annotations from convertMetadataDocumentV2
 * @param index - The annotation index to populate
 */
function processTargetV2Annotations(v2Annotations, index) {
    for (const v2Ann of v2Annotations) {
        if (v2Ann.target.kind !== "Property" /* Edm.Property */) {
            continue;
        }
        const uri = v2Ann.target.location?.uri ?? '';
        const propertyTarget = v2Ann.target.path;
        if (v2Ann.name === 'sap:text') {
            const textKey = buildAnnotationIndexKey(propertyTarget, constants_1.COMMON_TEXT);
            if (!index[textKey]) {
                const syntheticElement = createSyntheticElement("Path" /* Edm.Path */, v2Ann.value, v2Ann.valueRange);
                index[textKey] = {
                    undefined: {
                        source: uri,
                        target: propertyTarget,
                        term: constants_1.COMMON_TEXT,
                        top: { uri, value: syntheticElement },
                        layers: [{ uri, value: syntheticElement }]
                    }
                };
            }
        }
        else if (v2Ann.name === 'sap:label') {
            const labelKey = buildAnnotationIndexKey(propertyTarget, constants_1.COMMON_LABEL);
            if (!index[labelKey]) {
                const syntheticElement = createSyntheticElement("String" /* Edm.String */, v2Ann.value, v2Ann.valueRange);
                index[labelKey] = {
                    undefined: {
                        source: uri,
                        target: propertyTarget,
                        term: constants_1.COMMON_LABEL,
                        top: { uri, value: syntheticElement },
                        layers: [{ uri, value: syntheticElement }]
                    }
                };
            }
        }
    }
}
/**
 * Builds a service index from service artifacts.
 * Creates indexes for entity sets, entity containers, and annotations.
 * For OData V2 services, also injects synthetic index entries from
 * inline sap:text/sap:label attributes so that annotation rules can report on metadata.xml.
 *
 * @param artifacts - Service artifacts to index
 * @param documents - Document map to populate with annotation files
 * @param v2Annotations - Pre-parsed V2 annotations from convertMetadataDocumentV2
 * @returns Complete service index with entity sets and annotations
 */
function buildServiceIndex(artifacts, documents, v2Annotations = []) {
    const entitySets = {};
    for (const document of Object.values(artifacts.annotationFiles)) {
        documents[document.uri] = document;
    }
    let entityContainer;
    artifacts.metadataService.visitMetadataElements((element) => {
        // NOSONAR - TODO: check if we can handle CDS differences better
        if (element.kind === 'EntityContainer' || element.kind === 'service') {
            entityContainer = element;
        }
        else if (element.kind === 'EntitySet' || element.kind === 'entitySet') {
            entitySets[element.name] = element;
        }
    });
    const annotationIndex = indexAnnotationsByAnnotationPath(artifacts, v2Annotations);
    return {
        entitySets: entitySets,
        entityContainer,
        annotations: annotationIndex
    };
}
