// @sap/approuter/lib/utils/redis-store.js
'use strict';

const zlib = require('zlib');
const Redis = require('ioredis');
const tracer = require('../utils/logger').getTracer(__filename);
const getRedisClient = require('./redis-client-driver').getClient;
const REDIS_SCAN_KEYS_SIZE = 1000;
Redis.Promise = global.Promise;

const APPROUTER_SESSION_PREFIX = '{approuter-sessions}';
const EXTERNAL_SESSION_PREFIX = '{external-sessions}';

const redisKey = (key, prefix = APPROUTER_SESSION_PREFIX) => `${ prefix }:${ key }`;

const compress = (data) => {
  const zip = zlib.gzipSync(Buffer.from(JSON.stringify(data)), { level: zlib.constants.Z_BEST_SPEED });
  const buff = new Buffer.from(zip);
  return buff.toString('base64');
};

const decompress = (base64data) => {
  if (!base64data) {
    return null;
  }
  const buff = new Buffer(base64data, 'base64');
  let zipBuff = Buffer.from(buff);
  const jsons = zlib.gunzipSync(zipBuff).toString();
  return JSON.parse(jsons);
};

const log = {
  error(msg, ...args) {
    if (tracer.isEnabled('Error')) {
      tracer.error(msg, ...args);
    }
  },
  debug(msg, ...args) {
    if (tracer.isEnabled('debug')) {
      tracer.debug(msg, ...args);
    }
  }
};

// Lua CAS script:
//   KEYS[1]  = session key
//   ARGV[1]  = expected current compressed value (empty string means key must not exist)
//   ARGV[2]  = new compressed value
//   ARGV[3]  = TTL in milliseconds
// Returns 1 on success, 0 on conflict (key changed since last read).
// Works on both standalone Redis and Redis Cluster because all keys accessed
// by the script share the same {approuter-sessions} hash tag and therefore
// land in the same hash slot.
const CAS_SCRIPT = `
local current = redis.call('GET', KEYS[1])
if current == false then current = '' end
if current ~= ARGV[1] then
  return 0
end
redis.call('SET', KEYS[1], ARGV[2], 'PX', ARGV[3])
return 1
`;

class RedisStore {
  constructor(extSessionMgt) {
    this.extSessionMgt = extSessionMgt;
  }

  checkClient() {
    this.client = getRedisClient(this.extSessionMgt);
    return !(this.client === null || this.client.closing);
  }

  async all(prefix) {
    if (this.checkClient()) {
      try {
        const keys = await this.getKeysByPatternV2(redisKey('*', prefix), this.client);
        log.debug(`Got keys from redis: ${ JSON.stringify(keys) }`);
        let compressedSessions = keys && keys.length > 0 && await this.client.mget(keys);
        let decompressedSessions = [];
        compressedSessions && compressedSessions.forEach((session) => {
          decompressedSessions.push(decompress(session));
        });
        return decompressedSessions;
      } catch (err) {
        log.error(`Failed to get all sessions: ${ err.message }`);
        return [];
      }
    } else {
      tracer.warning('No redis client!!');
      return [];
    }
  }

  async clear() {
    if (this.checkClient()) {
      try {
        const keys = await this.getKeysByPatternV2(redisKey('*'), this.client);
        log.debug(`Got keys from redis: ${ JSON.stringify(keys) }`);
        await this.client.del(keys);
      } catch (err) {
        log.error(`Failed to clear all sessions: ${ err.message }`);
      }
    } else {
      tracer.warning('No redis client!!');
    }
  }

  async destroy(sessionId) {
    if (this.checkClient()) {
      try {
        log.debug(`Destroying session with id ${ sessionId }`);
        await this.client.del(redisKey(sessionId));
      } catch (err) {
        log.error(`Failed destroying session with id ${ sessionId }: ${ err.message }`);
      }
    } else {
      tracer.warning('No redis client!!');
    }
  }

  async get(sessionId) {
    if (this.checkClient()) {
      log.debug(`Getting session with id ${ sessionId }`);
      try {
        const session = await this.client.get(redisKey(sessionId));
        return decompress(session);
      } catch (err) {
        log.error(`Failed to get session with id ${ sessionId }: ${ err.message }`);
        return null;
      }
    } else {
      tracer.warning('No redis client!!');
      return null;
    }
  }

  attachAdditionalPropertiesToSession(session, req) {
    session.req = req;
    session.reload = (cb) => cb();
    session.save = function (cb) { cb && cb(); };
    session.touch = function (){};
    req.session = session;
  }

  async getExternalSession(id, req, prefix) {
    if (this.checkClient()) {
      log.debug(`Getting external session with id ${ id }`);
      try {
        const session = await this.client.get(redisKey(id, (prefix || EXTERNAL_SESSION_PREFIX)));
        const decompressedSession = decompress(session);
        if (req) {
          this.attachAdditionalPropertiesToSession(decompressedSession, req);
        }
        return decompressedSession;
      } catch (err) {
        log.debug(`Failed to get external session with id ${ id }: ${ err.message }`);
        return null;
      }
    } else {
      tracer.warning('No redis client!!');
      return null;
    }
  }

  async updateExternalSession(session, prefix, uniqueKey = false) {
    if (this.checkClient()) {
      try {
        log.debug(`Storing external session with id ${ session.externalSessionId }`);
        let sessionToSave = Object.assign({}, session);
        delete sessionToSave.req;
        const setArgs = [
          redisKey(sessionToSave.externalSessionId, (prefix || EXTERNAL_SESSION_PREFIX)),
          compress(sessionToSave),
          'PX',
          (sessionToSave.externalSessionExpiration * 1000) - Date.now()
        ];
        if (uniqueKey) {
          setArgs.push('NX');
        }
        const result = await this.client.set(...setArgs);
        return result === 'OK'; // Redis returns 'OK' on success
      } catch (err) {
        log.error(`Failed to set external session with id ${ session.externalSessionId }: ${ err.message }`);
        return false;
      }
    } else {
      tracer.warning('No redis client!!');
      return false;
    }
  }

  async destroyExternalSession(id, prefix) {
    if (this.checkClient()) {
      try {
        log.debug(`Destroying external session with id ${ id }`);
        await this.client.del(redisKey(id, (prefix || EXTERNAL_SESSION_PREFIX)));
      } catch (err) {
        log.error(`Failed destroying service-to-approuter session with id ${ id }: ${ err.message }`);
      }
    } else {
      tracer.warning('No redis client!!');
    }
  }

  async length() {
    if (this.checkClient()) {
      try {
        const keys = await this.getKeysByPatternV2(redisKey('*'), this.client);
        log.debug(`Keys length: ${ keys.length }`);
        return keys.length;
      } catch (err) {
        log.error(`Failed to get keys length: ${ err.message }`);
      }
    } else {
      tracer.warning('No redis client!!');
      return 0;
    }
  }

  async ttl(sessionId) {
    if (this.checkClient()) {
      try {
        log.debug(`Getting session ttl with id ${ sessionId }`);
        return await this.client.ttl(redisKey(sessionId));
      } catch (err) {
        log.error(`Failed to get ttl for session with id ${ sessionId }: ${ err.message }`);
        return 0;
      }
    } else {
      tracer.warning('No redis client!!');
      return 0;
    }
  }

  async set(sessionId, session, timeout, useOptimisticLocking = false, expectedValue = undefined) {
    if (!this.checkClient()) {
      tracer.warning('No redis client!!');
      return;
    }

    const key = redisKey(sessionId);
    const newCompressed = compress(session);

    if (!useOptimisticLocking) {
      // Plain SET: session-stickiness mode (fire-and-forget, no conflict detection needed)
      try {
        log.debug(`Storing session with id ${ sessionId }`);
        await this.client.set(key, newCompressed, 'PX', timeout * 60 * 1000);
      } catch (err) {
        log.error(`Failed to set a session with id ${ sessionId }: ${ err.message }`);
      }
      return;
    }

    // Optimistic locking via a single atomic Lua CAS script.
    // One EVAL round-trip, no duplicate connections, works on both standalone
    // and Redis Cluster (all keys share the {approuter-sessions} hash tag).
    // CAS baseline: `expectedValue` is the JSON string last written by storeSession()
    // (memstore keeps this.sessions[id] = that exact string and never mutates it in-place).
    // Both sides of the comparison are produced by the same JSON.stringify() + gzip pipeline,
    // so key ordering and compressed output are guaranteed to be byte-identical for the same
    // logical content — no spurious CAS failures from key-order divergence.
    const expected = (expectedValue !== undefined && expectedValue !== null)
      ? compress(expectedValue)
      : '';

    try {
      log.debug(`CAS storing session with id ${ sessionId }`);
      const result = await this.client.eval(CAS_SCRIPT, 1, key, expected, newCompressed, String(timeout * 60 * 1000));
      if (result === 0) {
        log.debug(`CAS conflict for session ${ sessionId } — written by another instance`);
        return false;
      }
      return true;
    } catch (err) {
      log.error(`Failed to CAS set session ${ sessionId }: ${ err.message }`);
      return false;
    }
  }

  async exists(sessionId) {
    if (this.checkClient()) {
      try {
        log.debug(`Checking existence of session with id ${ sessionId }`);
        const exists = await this.client.exists(redisKey(sessionId));
        log.debug(`Session with id ${ sessionId } exists ? ${ !!exists }`);
        return !!exists;
      } catch (err) {
        log.error(`Failed to check if session ${ sessionId } exists: ${ err.message }`);
        return false;
      }
    } else {
      tracer.warning('No redis client!!');
      return false;
    }
  }

  async resetTimer(sessionId, timeout) {
    if (this.checkClient()) {
      try {
        log.debug(`Resetting time of session with id ${ sessionId } to ${ timeout } minutes`);
        await this.client.expire(redisKey(sessionId), timeout * 60);
      } catch (err) {
        log.error(`Failed to expire session with id ${ sessionId }: ${ err.message }`);
      }
    } else {
      tracer.warning('No redis client!!');
    }
  }

  async getKeysByPatternV2(key, client) {
    return new Promise((resolve) => {
      // Select slave node for read scaling if configured, otherwise use master
      let node = client;
      if (client.options.scaleReads) {
        const slaveNodes = client.nodes('slave');
        if (slaveNodes && slaveNodes.length > 0) {
          // Use slave node for read operations (optimal)
          node = slaveNodes[0];
        } else {
          // Fallback to master node if no slave nodes available (e.g., during cluster topology changes)
          tracer.warning('Redis cluster has no slave nodes available, falling back to master node for read operations. This may indicate cluster scaling or failover in progress.');
        }
      }
      const stream = node.scanStream({
        match: key,
        count: REDIS_SCAN_KEYS_SIZE
      });
      const keysFound = [];
      stream.on('data', async function (resultKeys) {
        // remove duplicates if exist - due to SCAN's implementation in Redis.
        const keys = [...new Set(resultKeys)];
        keysFound.push(...keys);
      });
      stream.on('end', async function () {
        log.debug('getKeysByPatternV2 - keys found: %s', JSON.stringify(keysFound));
        resolve(keysFound);
      });

      stream.on('error', function (err) {
        log.error('getKeysByPatternV2 - failed to get keys by pattern %s, err %s', key, err.message);
        // mitigation flag - will not throw exception - avoid using
        resolve([]);
      });
    });
  }
}

let store = null;

module.exports.getStore = (extSessionMgt) => {
  if (!store) {
    store = new RedisStore(extSessionMgt);
  }
  return store;
};
