// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-text-arrangement-hidden.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const odata_annotation_core_1 = require("@sap-ux/odata-annotation-core");
const rule_factory_1 = require("../language/rule-factory");
const diagnostics_1 = require("../language/diagnostics");
const parser_1 = require("../project-context/parser");
const constants_1 = require("../constants");
const common_text_helpers_1 = require("./utils/common-text-helpers");
/**
 * Checks whether a Common.Text annotation element has a nested UI.TextArrangement inline annotation.
 *
 * @param textElement - The Common.Text annotation element
 * @param aliasInfo - Alias information for the file containing the annotation
 * @returns True if an inline UI.TextArrangement child annotation was found
 */
function hasInlineTextArrangement(textElement, aliasInfo) {
    if (!aliasInfo) {
        return false;
    }
    for (const child of textElement.content) {
        if (child.type !== odata_annotation_core_1.ELEMENT_TYPE) {
            continue;
        }
        const childElement = child;
        if (childElement.name !== "Annotation" /* Edm.Annotation */) {
            continue;
        }
        const childTerm = (0, odata_annotation_core_1.getElementAttributeValue)(childElement, "Term" /* Edm.Term */);
        if (!childTerm) {
            continue;
        }
        const qualifiedTerm = (0, odata_annotation_core_1.toFullyQualifiedName)(aliasInfo.aliasMap, aliasInfo.currentFileNamespace, (0, odata_annotation_core_1.parseIdentifier)(childTerm));
        if (qualifiedTerm === 'com.sap.vocabularies.UI.v1.TextArrangement') {
            return true;
        }
    }
    return false;
}
/**
 * Collects all entity type names that have UI.TextArrangement applied directly at entity-type level.
 *
 * @param parsedService - The parsed OData service
 * @returns Set of fully-qualified entity type names with entity-level UI.TextArrangement
 */
function collectEntityTypesWithTextArrangement(parsedService) {
    const entityTypes = new Set();
    for (const key of Object.keys(parsedService.index.annotations)) {
        const atIdx = key.indexOf('/@');
        if (atIdx === -1) {
            continue;
        }
        const targetPath = key.substring(0, atIdx);
        const term = key.substring(atIdx + 2);
        if (term === constants_1.UI_TEXT_ARRANGEMENT && !targetPath.includes('/')) {
            entityTypes.add(targetPath);
        }
    }
    return entityTypes;
}
/**
 * Checks whether UI.Hidden is set on the given property and produces diagnostics if so.
 *
 * @param textPropertyTarget - Fully-qualified target path of the text property (e.g. "Service.Entity/prop")
 * @param targetPath - The annotation target path that has Common.Text and UI.TextArrangement
 * @param pageNames - Page target names where the annotated entity type is used
 * @param parsedService - The parsed OData service
 * @returns Array of diagnostics (empty if the property is not hidden)
 */
function checkHiddenProperty(textPropertyTarget, targetPath, pageNames, parsedService) {
    const problems = [];
    const hiddenKey = (0, parser_1.buildAnnotationIndexKey)(textPropertyTarget, constants_1.UI_HIDDEN);
    const hiddenAnnotations = parsedService.index.annotations[hiddenKey];
    if (!hiddenAnnotations) {
        return problems;
    }
    for (const hiddenAnnotation of Object.values(hiddenAnnotations)) {
        // Skip only when explicitly set to false (Bool="false" means not hidden)
        // Dynamic path expressions (Path="...") are still warned — presence of
        // UI.Hidden on the text property is considered problematic regardless
        const boolVal = (0, odata_annotation_core_1.getElementAttributeValue)(hiddenAnnotation.top.value, "Bool" /* Edm.Bool */);
        if (boolVal === 'false') {
            continue;
        }
        problems.push({
            type: diagnostics_1.TEXT_ARRANGEMENT_HIDDEN,
            pageNames,
            annotation: {
                reference: hiddenAnnotation.top,
                textPropertyPath: textPropertyTarget,
                targetWithTextArrangement: targetPath
            }
        });
    }
    return problems;
}
/**
 * Processes a single Common.Text annotation and returns diagnostics if the referenced text property is hidden.
 *
 * @param textAnnotation - The indexed Common.Text annotation to process
 * @param entityTypeName - Fully-qualified name of the entity type that owns the annotated property
 * @param targetPath - The annotation target path (e.g. "Service.Entity/property")
 * @param pageNames - Page target names where the entity type is used
 * @param entityTypesWithTextArrangement - Entity types with UI.TextArrangement at entity-type level
 * @param parsedService - The parsed OData service
 * @returns Array of diagnostics (empty if no violation found)
 */
function processTextAnnotation(textAnnotation, entityTypeName, targetPath, pageNames, entityTypesWithTextArrangement, parsedService) {
    const textElement = textAnnotation.top.value;
    const textPath = (0, odata_annotation_core_1.getElementAttributeValue)(textElement, "Path" /* Edm.Path */);
    if (!textPath) {
        return [];
    }
    // UI.TextArrangement may be a nested inline annotation inside Common.Text
    // (property level, takes precedence) or applied directly on the entity type
    // (entity-type level fallback per vocabulary spec)
    const aliasInfo = parsedService.artifacts.aliasInfo[textAnnotation.top.uri];
    const hasTextArrangement = hasInlineTextArrangement(textElement, aliasInfo) || entityTypesWithTextArrangement.has(entityTypeName);
    if (!hasTextArrangement) {
        return [];
    }
    const resolved = (0, common_text_helpers_1.resolveTextPropertyPath)(entityTypeName, textPath, parsedService.artifacts.metadataService);
    if (!resolved) {
        return [];
    }
    const textPropertyTarget = `${resolved.entityTypeName}/${resolved.propertyName}`;
    return checkHiddenProperty(textPropertyTarget, targetPath, pageNames, parsedService);
}
/**
 * Processes a single annotation index entry and returns any diagnostics found.
 *
 * @param annotationKey - The annotation index key (e.g. "Service.Entity/prop/\@Common.Text")
 * @param qualifiedAnnotations - All qualified annotations for this key, keyed by qualifier
 * @param entityTypesWithTextArrangement - Entity types with UI.TextArrangement at entity-type level
 * @param parsedService - The parsed OData service
 * @param relevantEntityTypes - Map from entity type name to page target names where it is used
 * @returns Array of diagnostics (empty if no violation found)
 */
function processAnnotationEntry(annotationKey, qualifiedAnnotations, entityTypesWithTextArrangement, parsedService, relevantEntityTypes) {
    const parsed = (0, common_text_helpers_1.parseCommonTextAnnotationKey)(annotationKey, relevantEntityTypes);
    if (!parsed) {
        return [];
    }
    const { targetPath, entityTypeName, pageNames } = parsed;
    const problems = [];
    for (const textAnnotation of Object.values(qualifiedAnnotations)) {
        problems.push(...processTextAnnotation(textAnnotation, entityTypeName, targetPath, pageNames, entityTypesWithTextArrangement, parsedService));
    }
    return problems;
}
/**
 * Collects all TextArrangementHidden diagnostics for a single parsed OData service.
 *
 * @param parsedService - The parsed OData service to check
 * @param relevantEntityTypes - Map from entity type name to page target names where it is used
 * @returns Array of diagnostics found in the service
 */
function collectProblemsForService(parsedService, relevantEntityTypes) {
    // Pre-pass: collect entity types that have UI.TextArrangement applied directly
    // (entity-type level acts as a fallback for all Common.Text properties on that type)
    const entityTypesWithTextArrangement = collectEntityTypesWithTextArrangement(parsedService);
    const problems = [];
    for (const [annotationKey, qualifiedAnnotations] of Object.entries(parsedService.index.annotations)) {
        problems.push(...processAnnotationEntry(annotationKey, qualifiedAnnotations, entityTypesWithTextArrangement, parsedService, relevantEntityTypes));
    }
    return problems;
}
const rule = (0, rule_factory_1.createFioriRule)({
    ruleId: diagnostics_1.TEXT_ARRANGEMENT_HIDDEN,
    meta: {
        type: 'problem',
        docs: {
            recommended: true,
            description: 'The description (text) property referenced by a UI.TextArrangement annotation must not have UI.Hidden set to true',
            url: 'https://github.com/SAP/open-ux-tools/blob/main/packages/eslint-plugin-fiori-tools/docs/rules/sap-text-arrangement-hidden.md'
        },
        messages: {
            [diagnostics_1.TEXT_ARRANGEMENT_HIDDEN]: 'The text property "{{textPropertyPath}}" referenced using the Common.Text annotation on "{{targetPath}}" is hidden (UI.Hidden). Remove the UI.Hidden annotation from the text property or set it to false.'
        }
    },
    check(context) {
        const problems = [];
        for (const [appKey, app] of Object.entries(context.sourceCode.projectContext.linkedModel.apps)) {
            const parsedApp = context.sourceCode.projectContext.index.apps[appKey];
            const parsedService = context.sourceCode.projectContext.getIndexedServiceForMainService(parsedApp);
            if (!parsedService) {
                continue;
            }
            const relevantEntityTypes = (0, common_text_helpers_1.collectRelevantEntityTypes)(app.pages, parsedService);
            problems.push(...collectProblemsForService(parsedService, relevantEntityTypes));
        }
        return problems;
    },
    createAnnotations(context, validationResult) {
        if (validationResult.length === 0) {
            return {};
        }
        const lookup = new Map();
        for (const diagnostic of validationResult) {
            lookup.set(diagnostic.annotation.reference.value, diagnostic);
        }
        return {
            ['target>element[name="Annotation"]'](node) {
                const diagnostic = lookup.get(node);
                if (!diagnostic) {
                    return;
                }
                context.report({
                    node,
                    messageId: diagnostics_1.TEXT_ARRANGEMENT_HIDDEN,
                    data: {
                        textPropertyPath: diagnostic.annotation.textPropertyPath,
                        targetPath: diagnostic.annotation.targetWithTextArrangement
                    }
                });
            }
        };
    }
});
exports.default = rule;
