// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-copy-to-clipboard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const diagnostics_1 = require("../language/diagnostics");
const rule_factory_1 = require("../language/rule-factory");
const rule_fixer_1 = require("../language/rule-fixer");
const helpers_1 = require("../utils/helpers");
const rule = (0, rule_factory_1.createFioriRule)({
    ruleId: diagnostics_1.COPY_TO_CLIPBOARD,
    meta: {
        type: 'suggestion',
        docs: {
            recommended: true,
            description: 'Sets whether to enable or disable the copy to clipboard button in tables',
            url: 'https://github.com/SAP/open-ux-tools/blob/main/packages/eslint-plugin-fiori-tools/docs/rules/sap-copy-to-clipboard.md'
        },
        messages: {
            [diagnostics_1.COPY_TO_CLIPBOARD]: 'Copy To Clipboard in the {{sectionText}}table must be correctly configured. If not set, the "Copy" button is displayed'
        },
        fixable: 'code'
    },
    check(context) {
        const problems = [];
        for (const [appKey, app] of Object.entries(context.sourceCode.projectContext.linkedModel.apps)) {
            const parsedApp = context.sourceCode.projectContext.index.apps[appKey];
            const parsedService = context.sourceCode.projectContext.getIndexedServiceForMainService(parsedApp);
            if (!parsedService) {
                continue;
            }
            for (const page of app.pages) {
                problems.push(...(0, helpers_1.checkAppTablesConfiguration)(page, parsedApp, checkConfiguration));
            }
        }
        return problems;
    },
    createJsonVisitorHandler: (context, diagnostic, deepestPathResult) => function report(node) {
        context.report({
            node,
            messageId: diagnostics_1.COPY_TO_CLIPBOARD,
            data: {
                sectionText: diagnostic.pageSectionName ? `${diagnostic.pageSectionName} ` : ''
            },
            fix: (0, rule_fixer_1.createJsonFixer)({
                context,
                deepestPathResult,
                node,
                operation: 'delete'
            })
        });
    }
});
/**
 *
 * @param table
 * @returns
 */
function isV2Table(table) {
    return 'copy' in table.configuration;
}
/**
 *
 * @param page
 * @param table
 * @param parsedApp
 * @param problems
 * @param pageSectionName
 */
function checkConfiguration(page, table, parsedApp, problems, pageSectionName) {
    let config;
    let wrongValue = false;
    if (isV2Table(table)) {
        config = table.configuration.copy;
    }
    else {
        config = table.configuration.disableCopyToClipboard;
        wrongValue = true;
    }
    if (config?.valueInFile === wrongValue) {
        const copyIssue = {
            type: diagnostics_1.COPY_TO_CLIPBOARD,
            pageName: page.targetName,
            pageSectionName,
            manifest: {
                uri: parsedApp.manifest.manifestUri,
                object: parsedApp.manifestObject,
                propertyPath: config.configurationPath
            }
        };
        problems.push(copyIssue);
    }
}
exports.default = rule;
